src/Microsoft.Diagnostics.Runtime/src/Common/ClrArray.cs

```diff
@@ -0,0 +1,124 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+
+namespace Microsoft.Diagnostics.Runtime
+{
+    /// <summary>
+    /// Represents an array in the target process.
+    /// </summary>
+    public struct ClrArray : IEquatable<ClrArray>, IEquatable<ClrObject>
+    {
+        /// <summary>
+        /// The address of the object.
+        /// </summary>
+        public ulong Address { get; }
+
+        /// <summary>
+        /// The type of the object.
+        /// </summary>
+        public ClrType Type { get; }
+
+        private int _length;
+
+        /// <summary>
+        /// Returns the count of elements in this array.
+        /// </summary>
+        public int Length
+        {
+            get
+            {
+                if (_length == -1)
+                {
+                    _length = Type.ClrObjectHelpers.DataReader.Read<int>(Address + (ulong)IntPtr.Size);
+                }
+
+                return _length;
+            }
+        }
+
+        public ClrArray(ulong address, ClrType type)
+        {
+            Address = address;
+            Type = type;
+            // default uninitialized value for size. Will be lazy loaded.
+            _length = -1;
+        }
+
+        /// <summary>
+        /// Gets <paramref name="count"/> value elements from the array.
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="count"></param>
+        /// <returns></returns>
+        public T[]? GetContent<T>(int count) where T : unmanaged
+        {
+            if (count > this.Length)
+            {
+                throw new ArgumentException("Cannot read more elements than elements in array.");
+            }
+
+            return Type.GetArrayElementsValues<T>(Address, count);
+        }
+
+        /// <summary>
+        /// Determines if this instance and another specific <see cref="ClrArray" /> have the same value.
+        /// <para>Instances are considered equal when they have same <see cref="Address" />.</para>
+        /// </summary>
+        /// <param name="other">The <see cref="ClrArray" /> to compare to this instance.</param>
+        /// <returns><c>true</c> if the <see cref="Address" /> of the parameter is same as <see cref="Address" /> in this instance; <c>false</c> otherwise.</returns>
+        public override bool Equals(object? other)
+        {
```

Comment:
```suggestion
        {
            if (other is null)
                return false;
```

An early out for null may be better.

Reply:
Yeah, sure.